---
title: "NEUROHACK 2022 "
date: "<h4>README updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output:
  github_document
---

## Team 

![](figures/logo/team_logo.png) 

### Members  

```{r}
members <- readxl::read_excel("presentations/team_members.xlsx")
data.frame(members)
```

### Project title 

Predicting ALS drug targets using integrative multi-modal machine learning 

### Goals   

1. Identify molecular targets (genes and variants) of ALS (status and survival time).  
2. Identify potential therapeutics for ALS.  

### Workflow

1. Preprocess input datasets  
    + MND_ALS VCFs  
	+ External datasets  
2. Filter input features  
3. Train predictive model  
4. Extract key genes
5. Identify therapeutics

## Abstract  

[ALS background info] 
[our approach] 
[main findings] 
[future directions]

## Materials  

The following datatypes were prepared as numeric 
participant x feature matrices that could be fed into our machine learning model.


### Phenotypes 

Participant x phenotype data for each of the 20 ALS participants were encoded as numeric vectors and fed as input to the model. This currently only include sex, but can easily be expanded to other categorical or continuous traits as they become available. 


### [Retroviral insertions](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/HERVK_Insertions)  

Retroviral insertions were identified using the pipeline described [here](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/blob/main/data/HERVK_Insertions/readme.md). 


### [SNPs: variant-level](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/SNP_VCFs/by_variant)

### [SNPs: gene-level](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/SNP_VCFs/by_gene)

### [cS2G: gene-level](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/cS2G/by_gene)

Unlike the aforementioned gene-level scores (which mapped SNPs-to-genes using overlap/proximity-based annotations already included in the VCF files) these SNP-to-gene mappings were made using scores generated by the combined SNPs-to-genes (cS2G) model. cS2G integrates data from various sources (e.g. QTLs, chromatin interactions) to create 
more accurate SNP-to-gene mappings. We specifically used the predictions generated for all SNP positions in the UK Biobank (found [here](https://alkesgroup.broadinstitute.org/cS2G/cS2G_UKBB/)). 

See the associated [preprint](https://www.medrxiv.org/content/10.1101/2021.08.02.21261488v1) for more details, as well as the [data](https://alkesgroup.broadinstitute.org/cS2G/).
 
### [Indels: variant-level](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/Indel_VCFs/by_variant)

### [SVs: variant-level](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/SV_VCFs/by_variant)   

### [SVs: gene-level](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/SV_VCFs/by_gene)
   

## Methods  

### Dimensionality reduction model 

[PCA/autoencoder description here by @Hanz]

**Figure 1.**:  

![](figures/models/autoencoder.png)

### Classifier model 

Each input modality (datatype) was fed into an supervised fully-connected artificial neural network (ANN), such that each participant is a sample and each gene/variant/annotation is a feature. The inputs of the model are initially partitioned into separate "channels" of the model input, and are reduced in dimensionality by the subsequent layers. Next, the reduced representation from each modality are concatenated into a single vector (layer 3 in **Fig. 2** below). Finally, the data is further compressed to predict which phenotypic category each participant belongs to. 

The model is currently designed to provide categorical predictions for each sample: short-survival vs. long-survival, or ALS vs. control (depending on the data available). However, it can also easily be adapted to continuous phenotypic data (e.g. survival years, GWAS-derived polygenic risk score (PRS)). 

Once fully trained, the model can be interrogated to extract the most relevant features per modality. This allows us to generate ranked lists of genes/variants/annotations which can be used in the candidate therapeutics  prediction step. 

**Figure 2**: 

![](figures/models/classifer.png) 

### Therapeutics prediction 
 
Once the relative importance of each gene for predicting ALS survival have been identified, three complementary approaches will be used to identify candidate 
therapeutics for ALS: 1. virtual screening, 2. perturbation database queries, 3. literature mining. 

#### Virtual screening

[description by @Yizhou here]

#### Perturbation database queries 

#### Literature mining  


 
## Results 

## Conclusions  

## Future directions

<hr>

# File descriptions 
 
## [data/SNP_VCFs/](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/SNP_VCFs) 

### by_variant/

### SNP_VCFs.all.tsv.gz 

All 20 MND ALS dummy VCFs from *SNP_VCFs/* merged into one participant x variant matrix.

The matrix is filled with genotypes numerically encoded as detailed [here](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/blob/main/data/genotype_encodings.csv). 

### SNP_VCFs.merged.tsv.gz 

All 20 MND ALS dummy VCFs from *SNP_VCFs/* merged into a `data.table` in long-format. 

This is the data from which *SNP_VCFs.all.tsv.gz* is derived.  


## [data/HERVK_Insertions/](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/tree/main/data/HERVK_Insertions)  

### HERV_K_Insertions.txt 

A dummy file with HERV_K insertions for all 20 MND ALS dummy subjects. 


<hr> 

## References 

> van Rheenen, W., van der Spek, R.A.A., Bakker, M.K. et al. Common and rare variant association analyses in amyotrophic lateral sclerosis identify 15 risk loci with distinct genetic architectures and neuron-specific biology. Nat Genet 53, 1636â€“1648 (2021). [https://doi.org/10.1038/s41588-021-00973-1](https://doi.org/10.1038/s41588-021-00973-1) 


<hr> 

## Session info

<details>

```{r}
utils::sessionInfo()
```

</details>
<br> 