---
title: "query_VCF"
author: "Brian M. Schilder"
date: "11/01/2022"
output: html_document
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = here::here())
knitr::opts_knit$set(echo = TRUE, root.dir = here::here())

if(!require(remotes)) install.packages("remotes") 
 
source(here::here("code/gene_length.R"))
source(here::here("code/utils.R"))

library(dplyr)
library(ggplot2)

root <- "/Volumes/Steelix/research/MND_ALS"
```

# Import data

## Phenotypes 

```{r}
pheno <- readxl::read_excel("data/phenotype/pheno_data.xlsx")
data.table::fwrite(pheno,"data/phenotype/pheno_data.tsv", sep="\t")
```


## Import locus coordinates  
 
```{r}
topss <- data.table::fread("data/GWAS/tableS21_coords.csv") %>% 
    dplyr::mutate(CHR=paste0("chr",chromosome),
                  POS=base_pair_location)
```

## List VCFs

```{r}
snp_vcfs <- list.files(file.path(root,"SNP_VCFs"),".vcf$", 
                      full.names = TRUE, recursive = TRUE)
indel_vcfs <- list.files(file.path(root,"Indel_VCFs"),".vcf$", 
                      full.names = TRUE, recursive = TRUE)
sv_vcfs <- list.files(file.path(root,"SV_VCFs"),".vcf$", 
                      full.names = TRUE, recursive = TRUE)
```
 
# Query one VCF 

```{r}
vcf <- vcfs[1]
# Rsamtools::bgzip(vcf,overwrite = T)
# bgz <- paste0(vcf,".bgz")
# index_out <- VariantAnnotation::indexVcf(bgz)
# header <-  MungeSumstats:::read_header(bgz)
bp_distance <- 5e+05*1

out <- VariantAnnotation::readVcf(file = vcf)
out <- methods::as(methods::as(out,"VRanges"),"GRanges")


queries <- lapply(topss$Name, function(locus){
    message(locus)
    topss_sub <- subset(topss, Name==locus)
    dat <- rbind(topss_sub %>% 
                           dplyr::mutate(POS=topss_sub$POS - bp_distance),
                       topss_sub %>%
                           dplyr::mutate(POS=topss_sub$POS + bp_distance)) %>%
        dplyr::mutate(POS=as.integer(POS))
    # out2 <- echotabix::query_vcf(dat = dat[,.(SNP,POS,CHR)], 
    #                              vcf_url = index_out$path)
    # out <- methods::as(methods::as(out,"VRanges"),"GRanges")
    # out <- VariantAnnotation::scanVcf(file = vcf) 
    overlap <- echoannot::granges_overlap(dat1 = dat, 
                                          chrom_col.1 = "CHR",
                                          start_col.1 = "POS", 
                                          dat2 = out) %>%
        data.frame() %>% data.table::data.table()
     
    return(overlap)
    # 
    # if(nrow(out@fixed)>0){
    #     out_df <- MungeSumstats:::vcf2df(out)
    #     return(out_df)
    # } else {
    #     return(out)
    # } 
}) %>% `names<-`(topss$Name) %>%
    data.table::rbindlist(use.names = TRUE, idcol = "Locus")
```

# Merge VCF 

Merge gene-level scores using 

## Get metadata 

```{r, eval=FALSE}
vcf_list <- c(SNP=snp_vcfs[1], 
              SV=sv_vcfs[1], 
              Indel=indel_vcfs[1])

vcf_cols <- lapply(names(vcf_list), function(x){
    vcf_head <- VariantAnnotation::scanVcfHeader(vcf_list[[x]]) 
    cols <- rbind(VariantAnnotation::info(vcf_head),
                  VariantAnnotation::geno(vcf_head), 
                  VariantAnnotation::fixed(vcf_head)[["FILTER"]] %>%
                      data.frame() %>%
                      dplyr::mutate(Number=NA, Type=NA)) %>%
        data.frame() %>%
        dplyr::mutate(vcf_type=x) 
    cols <- cbind(Column=rownames(cols),cols)
    return(cols)
}) %>% do.call(what = "rbind") 
utils::write.csv(vcf_cols, "data/VCF_cols.csv", row.names = FALSE)
 

```



## SVs

Process Structural Variants (SVs) at the variant- and gene- levels. 

```{r}
all_ids <- vcfs2ids(vcfs = sv_vcfs)

#### Create variant-level matrices ####
genotypes_list <- genotype_matrix_grouped(vcfs = sv_vcfs, 
                                          grouping_var = "SVTYPE",
                                          save_prefix = "data/SV_VCFs/by_variant/SV_VCFs")

#### Create gene-level matrices ####
grl <- gather_vcfs(vcfs = sv_vcfs,
                   cols=c("id","SVTYPE","SVLEN","ensembl_gene_id","QUAL","GT"))
gr_dt <- merge_vcfs(grl = grl)
gene_counts <- get_gene_counts(gr_dt = gr_dt)
gene_counts <- add_gene_length(gene_counts = gene_counts)
gene_counts <- add_gene_score(gene_counts = gene_counts)
### Save matrices 
data_type <- "gene_scores" # gene_scores"
mat_list <- create_gene_matrix(gene_counts = gene_counts,
                               grouping_var = "SVTYPE",
                               data_type = data_type, 
                               save_prefix = "data/SV_VCFs/by_gene/SV_VCFs")
# mat <- do.call("cbind", mat_list)
# data.table::fwrite(gene_counts, "data/SV_VCFs/by_gene/SV_VCFs.gene_scores.long.tsv.gz")
```

### Participant similarity 

```{r} 
cor_mat_list <- create_corr_heatmaps(mat_list = mat_list,
                                     data_type = data_type)
```

### See whether cluster correspond to survival similarity

```{r} 
corr_pheno <-  cor_mat_list$DEL %>%
    data.table::data.table(keep.rownames = "iid") %>%
    data.table::melt.data.table(variable.name = "iid2", value.name = "corr") %>%
    dplyr::mutate_at(.vars = c("iid","iid2"),
                     .funs = function(x){gsub("__.*$","",x)}) %>%
    data.table::merge.data.table(data.table::data.table(pheno),
                                 by.x = "iid2",
                                 by.y = "iid") %>%
    subset(iid!=iid2) %>% 
    unique() 

ggplot(data=corr_pheno,
       aes(x=corr, y=Phe3_Surv_years, 
           color= PheFemale_Sex, size = Phe3_Surv_years)) +
    geom_point(alpha=.6) +
    geom_smooth(method = "lm") +
    ggpubr::stat_cor() +
    # ggpubr::stat_compare_means() +
    theme_bw()

```
```{r}

rsums <- Matrix::rowSums(mat_list$DEL)
sum_pheno <-  data.table::data.table(score=rsums,
                                     iid=names(rsums)) %>% 
    dplyr::mutate_at(.vars = c("iid"),
                     .funs = function(x){gsub("__.*$","",x)}) %>%
    data.table::merge.data.table(data.table::data.table(pheno),
                                 by = "iid") %>% 
    unique()


ggplot(data=sum_pheno,
       aes(x=score, y=Phe3_Surv_years, 
           color= PheFemale_Sex, size = Phe3_Surv_years)) +
    geom_point(alpha=.6) +
    geom_smooth(method = "lm") +
    ggpubr::stat_cor() +
    # ggpubr::stat_compare_means() +
    theme_bw()
```


## Indels 
 
```{r}
all_ids <- vcfs2ids(vcfs = indel_vcfs)

# vcf=VariantAnnotation::readVcf(indel_vcfs[1])

#### Create variant-level matrices ####
mat <- genotype_matrix(vcfs = indel_vcfs, 
                       structural_variants = TRUE,
                       save_prefix = "data/Indel_VCFs/by_variant/Indel_VCFs")

#### Create gene-level matrices ####
grl <- gather_vcfs(vcfs = indel_vcfs)
gr_dt <- merge_vcfs(grl = grl, 
                    save_prefix ="data/Indel_VCFs/by_variant/Indel_VCFs")
gene_counts <- get_gene_counts(gr_dt = gr_dt, 
                               gene_col = "CSQ", 
                               extract_string = "ENSG")
gene_counts <- add_gene_length(gene_counts = gene_counts)
gene_counts <- add_gene_score(gene_counts = gene_counts)
### Save matrices 
data_type <- "gene_scores" # gene_scores"
mat_list <- create_gene_matrix(gene_counts = gene_counts,
                               grouping_var = "SVTYPE",
                               data_type = data_type, 
                               save_prefix = "data/Indel_VCFs/by_gene/Indel_VCFs")
# mat <- do.call("cbind", mat_list)
# data.table::fwrite(gene_counts, "data/SV_VCFs/by_gene/SV_VCFs.gene_scores.long.tsv.gz")
```



## SNPs

```{r}
mat_var <- genotype_matrix(vcfs = snp_vcfs,
                       save_prefix="data/SNP_VCFs/by_variant/SNP_VCFs")

grl <- gather_vcfs(vcfs = snp_vcfs)
gr_dt <- merge_vcfs(grl = grl, 
                    save_prefix = "data/SNP_VCFs/by_variant/SNP_VCFs")
gene_counts <- get_gene_counts(gr_dt = gr_dt, 
                               gene_col = "CSQ", 
                               extract_string = "ENSG")
gene_counts <- add_gene_length(gene_counts = gene_counts)
gene_counts <- add_gene_score(gene_counts = gene_counts,
                              formula_str = "GT_int ~ count + TXLENGTH + QUAL")

data.table::fwrite(gene_counts, "data/SNP_VCFs/by_gene/SNP_VCFs.gene_scores.long.tsv.gz")

### Save matrices 
data_type <- "gene_scores" # gene_scores"
mat_gene <- create_gene_matrix(gene_counts = gene_counts, 
                               data_type = data_type, 
                               save_prefix = "data/SNP_VCFs/by_gene/SNP_VCFs")
```


<hr> 

# Session info

<details>

```{r}
utils::sessionInfo()
```

</details>
<br> 
