---
title: "query_VCF"
author: "Brian M. Schilder"
date: "11/01/2022"
output: html_document
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE, root.dir = here::here())
knitr::opts_knit$set(echo = TRUE, root.dir = here::here())

if(!require(remotes)) install.packages("remotes")
if(!require(echotabix)) remotes::install_github("RajLabMSSM/echotabix")
 
source(here::here("code/gene_length.R"))

library(dplyr)
library(ggplot2)

root <- "/Volumes/Steelix/research/MND_ALS"
```

# Import data

## Phenotypes 

```{r}
pheno <- readxl::read_excel("data/phenotype/pheno_data.xlsx")
```


## Import locus coordinates  
 
```{r}
topss <- data.table::fread("data/GWAS/tableS21_coords.csv") %>% 
    dplyr::mutate(CHR=paste0("chr",chromosome),
                  POS=base_pair_location)
```

## List VCFs

```{r}
snp_vcfs <- list.files(file.path(root,"SNP_VCFs"),".vcf$", 
                      full.names = TRUE, recursive = TRUE)
indel_vcfs <- list.files(file.path(root,"Indel_VCFs"),".vcf$", 
                      full.names = TRUE, recursive = TRUE)
sv_vcfs <- list.files(file.path(root,"SV_VCFs"),".vcf$", 
                      full.names = TRUE, recursive = TRUE)
```
 
# Query one VCF 

```{r}
vcf <- vcfs[1]
# Rsamtools::bgzip(vcf,overwrite = T)
# bgz <- paste0(vcf,".bgz")
# index_out <- VariantAnnotation::indexVcf(bgz)
# header <-  MungeSumstats:::read_header(bgz)
bp_distance <- 5e+05*1

out <- VariantAnnotation::readVcf(file = vcf)
out <- methods::as(methods::as(out,"VRanges"),"GRanges")


queries <- lapply(topss$Name, function(locus){
    message(locus)
    topss_sub <- subset(topss, Name==locus)
    dat <- rbind(topss_sub %>% 
                           dplyr::mutate(POS=topss_sub$POS - bp_distance),
                       topss_sub %>%
                           dplyr::mutate(POS=topss_sub$POS + bp_distance)) %>%
        dplyr::mutate(POS=as.integer(POS))
    # out2 <- echotabix::query_vcf(dat = dat[,.(SNP,POS,CHR)], 
    #                              vcf_url = index_out$path)
    # out <- methods::as(methods::as(out,"VRanges"),"GRanges")
    # out <- VariantAnnotation::scanVcf(file = vcf) 
    overlap <- echoannot::granges_overlap(dat1 = dat, 
                                          chrom_col.1 = "CHR",
                                          start_col.1 = "POS", 
                                          dat2 = out) %>%
        data.frame() %>% data.table::data.table()
     
    return(overlap)
    # 
    # if(nrow(out@fixed)>0){
    #     out_df <- MungeSumstats:::vcf2df(out)
    #     return(out_df)
    # } else {
    #     return(out)
    # } 
}) %>% `names<-`(topss$Name) %>%
    data.table::rbindlist(use.names = TRUE, idcol = "Locus")
```

# Merge VCF 

Merge gene-level scores using 

## Get metadata 

```{r, eval=FALSE}
vcf_list <- c(SNP=snp_vcfs[1], 
              SV=sv_vcfs[1], 
              Indel=indel_vcfs[1])

vcf_cols <- lapply(names(vcf_list), function(x){
    vcf_head <- VariantAnnotation::scanVcfHeader(vcf_list[[x]]) 
    cols <- rbind(VariantAnnotation::info(vcf_head),
                  VariantAnnotation::geno(vcf_head), 
                  VariantAnnotation::fixed(vcf_head)[["FILTER"]] %>%
                      data.frame() %>%
                      dplyr::mutate(Number=NA, Type=NA)) %>%
        data.frame() %>%
        dplyr::mutate(vcf_type=x) 
    cols <- cbind(Column=rownames(cols),cols)
    return(cols)
}) %>% do.call(what = "rbind") 
utils::write.csv(vcf_cols, "data/VCF_cols.csv", row.names = FALSE)
 

```



## Structural Variants (SV)

```{r}
all_ids <- vcfs2ids(vcfs = sv_vcfs)

grl <- gather_vcfs(vcfs = sv_vcfs,
                   cols=c("id","SVTYPE","SVLEN","ensembl_gene_id","QUAL","GT"))
genotypes_list <- genotype_matrices(
    grl = grl,
    grouping_var = "SVTYPE",
    save_prefix="data/SV_VCFs/by_variant/SV_VCFs")  

gr_dt <- merge_vcfs(grl = grl)
gene_counts <- get_gene_counts(gr_dt = gr_dt)

sv_gene_counts <- add_gene_length(gene_counts = gene_counts)
sv_gene_counts <- add_gene_score(sv_gene_counts = sv_gene_counts)
 

data_type <- "gene_genotypes"
# data_type <- "gene_scores"
types <- unique(sv_gene_counts$SVTYPE)
remove_nonvarying <- TRUE
mat_list <- lapply(types, function(type){ 
    M <- data.table::dcast.data.table(
        data = subset(sv_gene_counts, SVTYPE==type), 
        # formula = id ~ SVTYPE,
        formula = id__type ~ ensembl_gene_id,
        fun.aggregate = "mean",
        fill = 0,
        drop = FALSE,
        value.var = if(data_type=="gene_scores")"score"else{"GT_int"}, 
        na.rm=TRUE) %>%
    tibble::column_to_rownames("id__type") %>%
    as.matrix() %>%
    as("sparseMatrix")
    #### Remove features that don't vary across participants ####
    if(remove_nonvarying){
        M <- M[,Matrix::colSums(M, na.rm = TRUE)>0]
    }
    message(type," : ",paste(dim(M),collapse = " x "))
    return(M)
}) %>% `names<-`(types)

# mat <- do.call("cbind", mat_list)

saveRDS(mat_list, paste0("data/SV_VCFs/SV_VCFs.",data_type,".rds"))

# mat_list <- readRDS( "data/SV_SNPs/SV_SNPs.gene_scores.rds")
# saveRDS(mat, "data/SV_SNPs/SV_SNPs.participant_scores.rds")
data.table::fwrite(sv_gene_counts, "data/SV_SNPs/SV_VCFs.gene_scores.long.tsv.gz")
```

## Check how similar individuals are

```{r} 
cor_mat_list <- lapply(names(mat_list), function(x){
    #### corr ####
    mat <- mat_list[[x]]
    cor_mat <- cor(as.matrix(Matrix::t(mat)))
    cor_mat[is.na(cor_mat)] <- 0
    diag(cor_mat) <- NA
    ### plot ###
    path <- paste0("figures/heatmaps/SV_",x,".",data_type,".pdf")
    pdf(path)
    heatmap(cor_mat, symm = TRUE)
    dev.off()
    cor_mat
}) %>% `names<-`(names(mat_list))

```

## See whether cluster correspond to surival similarity

```{r} 
corr_pheno <-  cor_mat_list$DEL %>%
    data.table::data.table(keep.rownames = "iid") %>%
    data.table::melt.data.table(variable.name = "iid2", value.name = "corr") %>%
    dplyr::mutate_at(.vars = c("iid","iid2"),
                     .funs = function(x){gsub("__.*$","",x)}) %>%
    data.table::merge.data.table(data.table::data.table(pheno),
                                 by.x = "iid2",
                                 by.y = "iid") %>%
    subset(iid!=iid2) %>% 
    unique() 

ggplot(data=corr_pheno,
       aes(x=corr, y=Phe3_Surv_years, 
           color= PheFemale_Sex, size = Phe3_Surv_years)) +
    geom_point(alpha=.6) +
    geom_smooth(method = "lm") +
    ggpubr::stat_cor() +
    # ggpubr::stat_compare_means() +
    theme_bw()

```
```{r}

rsums <- Matrix::rowSums(mat_list$INV)
sum_pheno <-  data.table::data.table(score=rsums,
                                     iid=names(rsums)) %>% 
    dplyr::mutate_at(.vars = c("iid"),
                     .funs = function(x){gsub("__.*$","",x)}) %>%
    data.table::merge.data.table(data.table::data.table(pheno),
                                 by = "iid") %>% 
    unique()


ggplot(data=sum_pheno,
       aes(x=score, y=Phe3_Surv_years, 
           color= PheFemale_Sex, size = Phe3_Surv_years)) +
    geom_point(alpha=.6) +
    geom_smooth(method = "lm") +
    ggpubr::stat_cor() +
    # ggpubr::stat_compare_means() +
    theme_bw()
```


## SNPs

```{r}
all_ids <- vcfs2ids(vcfs = snp_vcfs)

grl <- gather_vcfs(vcfs = snp_vcfs)
genotypes_list <- genotype_matrices(
    grl = grl,
    grouping_var = NULL,
    save_prefix="data/SNP_VCFs/by_variant/SNP_VCFs")  

gr_dt <- merge_vcfs(grl = grl)
gene_counts <- get_gene_counts(gr_dt = gr_dt)

sv_gene_counts <- add_gene_length(gene_counts = gene_counts)
sv_gene_counts <- add_gene_score(sv_gene_counts = sv_gene_counts)
 
```


<hr> 

# Session info

<details>

```{r}
utils::sessionInfo()
```

</details>
<br> 
