# *model_input/* file descriptions 

## Phenotypes 

### pheno_data.tsv 

Participant x phenotype data for each of the 20 ALS participants.  

Column descriptions:  
*iid*: Participant ID.
*PheFemale_Sex*: Sex. 
*Phe3_Surv_years*: Survival years. 

**Creator**: Brian Schilder


## Retroviral insertions 

### HERV_K_Insertions.txt 

A participant x region dummy file with HERV_K insertions for all 20 MND ALS dummy subjects.  

[Additional info.](https://github.com/DEMON-NEUROHACK/Challenge-3-London-Team-C/blob/main/data/HERVK_Insertions/readme.md)

**Creator**: Renata Kabiljo 


## SNPs: variant-level

### SNP_VCFs.genotype_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *SNP_VCFs/* merged into one participant x variant matrix. 

**Creator**: Brian Schilder


## SNPs: gene-level 

Variant-to-gene mappings were estimated by aggregating gene annotations included in each VCF file (which are primarily based on proximity and overlap). 

### SNP_VCFs.gene_scores_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *SNP_VCFs/* merged into one participant x gene matrix. 

**Creator**: Brian Schilder 

## cS2G: gene-level 

### SNP_VCFs.cS2G_scores_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *SNP_VCFs/* merged into one participant x gene matrix.
Variant-to-gene mappings performed using predictions from the [combined SNPs-to-genes (cS2G) model](https://www.medrxiv.org/content/10.1101/2021.08.02.21261488v1).  

**Creator**: Brian Schilder 

## Indels: variant-level 

Variant-to-gene mappings were estimated by aggregating gene annotations included in each VCF file (which are primarily based on proximity and overlap). 

### Indel_VCFs.genotype_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *Indel_VCFs/* merged into one participant x variant matrix.  

**Creator**: Brian Schilder

### Indel_VCFs.gene_scores_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *Indel_VCFs/* merged into one participant x gene matrix.  

**Creator**: Brian Schilder


## SVs: variant-level

### SV_VCFs.DEL.genotype_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x variant matrix. Deletion SVs only.  

**Creator**: Brian Schilder

### SV_VCFs.DUP-TANDEM.genotype_matrix.tsv.gz

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x variant matrix. Duplications/tandem repeat SVs only.  

**Creator**: Brian Schilder

### SV_VCFs.INS.genotype_matrix.tsv.gz

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x variant matrix. Insertion SVs only.  

**Creator**: Brian Schilder

### SV_VCFs.INV.genotype_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x variant matrix. Inversion SVs only.   

**Creator**: Brian Schilder 


## SVs: gene-level  

Variant-to-gene mappings were estimated by aggregating gene annotations included in each VCF file (which are primarily based on proximity and overlap). 

### SV_VCFs.DEL.gene_scores_matrix.tsv.gz  

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x gene matrix. Deletion SVs only.   

**Creator**: Brian Schilder  

### SV_VCFs.DUP-TANDEM.gene_scores_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x gene matrix. Duplication/tandem repeat SVs only.   

**Creator**: Brian Schilder  

### SV_VCFs.INS.gene_scores_matrix.tsv.gz  

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x gene matrix. Insertion SVs only.   

**Creator**: Brian Schilder  

### SV_VCFs.INV.gene_scores_matrix.tsv.gz  

All 20 MND ALS dummy VCFs from *SV_VCFs/* merged into one participant x gene matrix. Inversion SVs only.   

**Creator**: Brian Schilder  

## cS2Gs: gene-level 

### SNP_VCFs.cS2G_scores_matrix.tsv.gz 

All 20 MND ALS dummy VCFs from *SNP_VCFs/* merged into one participant x gene matrix. 
However, unlike *SNP_VCFs.gene_scores_matrix.tsv.gz* (which mapped SNPs-to-genes using overlap/proximity-based annotations already included in the VCF files) these SNP-to-gene mappings were made using scores generated by the combined SNPs-to-genes (cS2G) model. cS2G integrates data from various sources (e.g. QTLs, chromatin interactions) to create 
more accurate SNP-to-gene mappings. We specifically used the predictions generated for all SNP positions in the UK Biobank (found [here](https://alkesgroup.broadinstitute.org/cS2G/cS2G_UKBB/)). 

See the associated [preprint](https://www.medrxiv.org/content/10.1101/2021.08.02.21261488v1) for more details, as well as the [data](https://alkesgroup.broadinstitute.org/cS2G/).


